File: android/Gutenberg/src/main/java/org/wordpress/gutenberg/stores/EditorURLCache.kt
package org.wordpress.gutenberg.stores

import android.annotation.SuppressLint
import android.util.Log
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.wordpress.gutenberg.model.EditorCachePolicy
import org.wordpress.gutenberg.model.http.EditorHTTPHeaders
import org.wordpress.gutenberg.model.http.EditorHttpMethod
import org.wordpress.gutenberg.model.http.EditorURLResponse
import java.io.File
import java.security.MessageDigest
import java.util.Date
import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.read
import kotlin.concurrent.write

/**
 * A URL-based cache for storing HTTP responses keyed by URL and HTTP method.
 *
 * Responses are stored on disk and survive process termination. Responses are keyed by both
 * URL and HTTP method, so GET and OPTIONS requests to the same URL are stored independently.
 */
class EditorURLCache(
    private val cacheRoot: File,
    private val cachePolicy: EditorCachePolicy = EditorCachePolicy.Always
) {
    private val json = Json { ignoreUnknownKeys = true }
    private val lock = ReentrantReadWriteLock()

    companion object {
        private const val TAG = "EditorURLCache"
    }

    init {
        cacheRoot.mkdirs()
    }

    /**
     * Stores a response for the given URL and HTTP method.
     *
     * If a response already exists for this URL and method combination, it will be overwritten.
     *
     * @param response The response to store.
     * @param url The URL to associate with the response.
     * @param httpMethod The HTTP method to associate with the response.
     */
    fun store(response: EditorURLResponse, url: String, httpMethod: EditorHttpMethod) {
        store(response, url, httpMethod, Date())
    }

    internal fun store(
        response: EditorURLResponse,
        url: String,
        httpMethod: EditorHttpMethod,
        currentDate: Date
    ) = lock.write {
        __store(response, url, httpMethod, currentDate)
    }

    /**
     * Stores the contents of a file as a cached response for the given URL and HTTP method.
     *
     * @param file The file whose contents should be stored.
     * @param headers The HTTP headers to associate with the response.
     * @param url The URL to associate with the response.
     * @param httpMethod The HTTP method to associate with the response.
     */
    fun store(
        file: File,
        headers: EditorHTTPHeaders,
        url: String,
        httpMethod: EditorHttpMethod
    ) {
        store(file, headers, url, httpMethod, Date())
    }

    internal fun store(
        file: File,
        headers: EditorHTTPHeaders,
        url: String,
        httpMethod: EditorHttpMethod,
        currentDate: Date
    ) = lock.write {
        val data = file.readText()
        val response = EditorURLResponse(data = data, responseHeaders = headers)
        __store(response, url, httpMethod, currentDate)
    }

    private fun __store(
        response: EditorURLResponse,
        url: String,
        httpMethod: EditorHttpMethod,
        currentDate: Date
    ) {
        val cacheKey = buildCacheKey(url, httpMethod)
        val cacheFile = File(cacheRoot, cacheKey)

        val entry = CachedEntry(
            data = response.data,
            headers = response.responseHeaders.dictionaryValue,
            storageDate = currentDate.time
        )

        val encodedEntry = json.encodeToString(entry)
        cacheFile.writeText(encodedEntry)
        Log.d(TAG, "Wrote cache entry: file=${cacheFile.absolutePath}, size=${encodedEntry.length} bytes, url=$url")
    }

    /**
     * Checks whether a cached response exists for the given URL and HTTP method.
     *
     * @param url The URL to check.
     * @param httpMethod The HTTP method to check.
     * @return `true` if a cached response exists, `false` otherwise.
     */
    fun hasResponse(url: String, httpMethod: EditorHttpMethod): Boolean {
        return hasResponse(url, httpMethod, Date())
    }

    internal fun hasResponse(url: String, httpMethod: EditorHttpMethod, currentDate: Date): Boolean {
        return getResponse(url, httpMethod, currentDate) != null
    }

    /**
     * Retrieves the cached response for the given URL and HTTP method.
     *
     * @param url The URL to look up.
     * @param httpMethod The HTTP method to look up.
     * @return The cached response, or `null` if no response is cached or the cache has expired.
     */
    fun getResponse(url: String, httpMethod: EditorHttpMethod): EditorURLResponse? {
        return getResponse(url, httpMethod, Date())
    }

    internal fun getResponse(
        url: String,
        httpMethod: EditorHttpMethod,
        currentDate: Date
    ): EditorURLResponse? = lock.read {
        val cacheKey = buildCacheKey(url, httpMethod)
        val cacheFile = File(cacheRoot, cacheKey)

        if (!cacheFile.exists()) {
            return@read null
        }

        runCatching {
            val entry = json.decodeFromString<CachedEntry>(cacheFile.readText())
            val storageDate = Date(entry.storageDate)

            if (!cachePolicy.allowsResponseWith(storageDate, currentDate)) {
                return@read null
            }

            EditorURLResponse(
                data = entry.data,
                responseHeaders = EditorHTTPHeaders(entry.headers)
            )
        }.getOrNull()
    }

    /**
     * Removes all cached responses.
     */
    fun purge() = lock.write {
        cacheRoot.listFiles()?.forEach { it.delete() }
    }

    /**
     * Removes cached responses that are no longer valid according to the cache policy.
     *
     * This method iterates through all cached entries and deletes those whose storage date
     * is no longer allowed by the current cache policy. Use this periodically to prevent
     * unbounded cache growth.
     *
     * @return The number of entries removed.
     */
    fun clean(): Int {
        return clean(Date())
    }

    internal fun clean(currentDate: Date): Int = lock.write {
        var removedCount = 0

        cacheRoot.listFiles()?.forEach { file ->
            if (!file.isFile) return@forEach

            val shouldRemove = runCatching {
                val entry = json.decodeFromString<CachedEntry>(file.readText())
                val storageDate = Date(entry.storageDate)
                !cachePolicy.allowsResponseWith(storageDate, currentDate)
            }.getOrDefault(true) // Remove entries that can't be parsed

            if (shouldRemove) {
                if (file.delete()) {
                    removedCount++
                    Log.d(TAG, "Cleaned expired cache entry: ${file.name}")
                }
            }
        }

        if (removedCount > 0) {
            Log.d(TAG, "Cache cleanup complete: removed $removedCount entries")
        }

        removedCount
    }

    /**
     * Builds a unique cache key for the given URL and HTTP method combination.
     */
    private fun buildCacheKey(url: String, httpMethod: EditorHttpMethod): String {
        val input = "$httpMethod:$url"
        val digest = MessageDigest.getInstance("SHA-256")
        val hashBytes = digest.digest(input.toByteArray())
        return hashBytes.joinToString("") { "%02x".format(it) }
    }

    /**
     * Internal structure for persisting cached responses.
     */
    @SuppressLint("UnsafeOptInUsageError")
    @Serializable
    private data class CachedEntry(
        val data: String,
        val headers: Map<String, String>,
        val storageDate: Long
    )
}

File: ios/Sources/GutenbergKit/Sources/Stores/EditorURLCache.swift
import Foundation
import OSLog

/// A URL-based cache for storing HTTP responses keyed by URL and HTTP method.
///
/// Responses are stored on disk and survive process termination. Responses are keyed by both URL and HTTP method, so GET and OPTIONS requests to the
/// same URL are stored independently.
///
/// Backed by `SQLiteKVCache`. The "one instance per backing file" contract from
/// `SQLiteKVCache` carries over: two `EditorURLCache` instances pointed at the
/// same `cacheRoot` is undefined behavior.
public struct EditorURLCache: Sendable {
    /// About enough for 10 sites of cached responses.
    private static let diskCapacity = 100 * 1024 * 1024

    private let store: SQLiteKVCache
    private let cachePolicy: EditorCachePolicy
    private let performanceMonitor = SignpostMonitor(for: Logger.performance)

    /// Creates a new URL cache.
    ///
    /// - Parameters:
    ///   - cacheRoot: The directory where cached responses will be stored.
    ///     If `nil`, the system default cache directory is used. The cache has a
    ///     maximum disk capacity of 100 MB.
    ///   - cachePolicy: The policy that determines when cached responses are considered valid.
    ///     Use `.ignore` to always fetch fresh data, `.maxAge(_:)` to expire entries after
    ///     a time interval, or `.always` (the default) to use cached data regardless of age.
    public init(cacheRoot: URL? = nil, cachePolicy: EditorCachePolicy = .always) {
        self.store = SQLiteKVCache(
            handle: "editorurlcache",
            directory: cacheRoot ?? URL.cachesDirectory,
            diskCapacity: Self.diskCapacity
        )
        self.cachePolicy = cachePolicy
    }

    /// Stores a response for the given URL and HTTP method.
    ///
    /// If a response already exists for this URL and method combination, it will be overwritten.
    ///
    /// - Parameters:
    ///   - response: The response to store.
    ///   - url: The URL to associate with the response.
    ///   - httpMethod: The HTTP method to associate with the response.
    /// - Throws: An error if the response cannot be stored.
    public func store(_ response: EditorURLResponse, for url: URL, httpMethod: EditorHttpMethod) throws {
        try self.store(response, for: url, httpMethod: httpMethod, currentDate: .now)
    }

    func store(
        _ response: EditorURLResponse,
        for url: URL,
        httpMethod: EditorHttpMethod,
        currentDate: Date
    ) throws {
        try self.store.put(
            key: Self.key(url: url, httpMethod: httpMethod),
            storageDate: currentDate,
            metadata: try JSONEncoder().encode(response.responseHeaders),
            value: response.data
        )
    }

    /// Stores the contents of a downloaded file as a cached response for the given URL and HTTP method.
    ///
    /// If a response already exists for this URL and method combination, it will be overwritten.
    ///
    /// - Parameters:
    ///   - path: The file URL whose contents should be stored.
    ///   - headers: The HTTP headers to associate with the response.
    ///   - url: The URL to associate with the response.
    ///   - httpMethod: The HTTP method to associate with the response.
    /// - Throws: An error if the file cannot be read or the response cannot be stored.
    public func store(
        fileAt path: URL,
        headers: EditorHTTPHeaders,
        for url: URL,
        httpMethod: EditorHttpMethod
    ) throws {
        try self.store(fileAt: path, headers: headers, for: url, httpMethod: httpMethod, currentDate: .now)
    }

    func store(
        fileAt path: URL,
        headers: EditorHTTPHeaders,
        for url: URL,
        httpMethod: EditorHttpMethod,
        currentDate: Date
    ) throws {
        try self.store.put(
            key: Self.key(url: url, httpMethod: httpMethod),
            storageDate: currentDate,
            metadata: try JSONEncoder().encode(headers),
            value: try Data(contentsOf: path)
        )
    }

    /// Checks whether a cached response exists for the given URL and HTTP method.
    ///
    /// - Parameters:
    ///   - url: The URL to check.
    ///   - httpMethod: The HTTP method to check.
    /// - Returns: `true` if a cached response exists, `false` otherwise.
    /// - Throws: An error if the check cannot be performed.
    public func hasData(for url: URL, httpMethod: EditorHttpMethod) throws -> Bool {
        try self.response(for: url, httpMethod: httpMethod, currentDate: .now) != nil
    }

    func hasData(for url: URL, httpMethod: EditorHttpMethod, currentDate: Date) throws -> Bool {
        try self.response(for: url, httpMethod: httpMethod, currentDate: currentDate) != nil
    }

    /// Retrieves the cached response for the given URL and HTTP method.
    ///
    /// - Parameters:
    ///   - url: The URL to look up.
    ///   - httpMethod: The HTTP method to look up.
    /// - Returns: The cached response, or `nil` if no response is cached.
    /// - Throws: An error if the response cannot be retrieved.
    public func response(for url: URL, httpMethod: EditorHttpMethod) throws -> EditorURLResponse? {
        try self.response(for: url, httpMethod: httpMethod, currentDate: .now)
    }

    func response(
        for url: URL,
        httpMethod: EditorHttpMethod,
        currentDate: Date
    ) throws -> EditorURLResponse? {
        try performanceMonitor.measure { () throws -> EditorURLResponse? in
            guard
                let entry = try self.store.get(key: Self.key(url: url, httpMethod: httpMethod)),
                self.cachePolicy.allowsResponseWith(date: entry.storageDate, currentDate: currentDate)
            else {
                return nil
            }
            let headers = try JSONDecoder().decode(EditorHTTPHeaders.self, from: entry.metadata)
            return EditorURLResponse(data: entry.value, responseHeaders: headers)
        }
    }

    /// Removes all cached responses.
    ///
    /// - Throws: An error if the cache cannot be cleared.
    public func clear() throws {
        try self.store.clear()
    }

    /// Combines the HTTP method and URL into a single string key. `SQLiteKVCache`
    /// hashes the key with SHA-256 before binding to SQLite, so length, escaping,
    /// and encoding aren't concerns here.
    private static func key(url: URL, httpMethod: EditorHttpMethod) -> String {
        "\(httpMethod.rawValue) \(url.absoluteString)"
    }
}
